stl/inc/yvals_core.h

```diff
@@ -306,6 +306,16 @@
 // _HAS_NODISCARD (in vcruntime.h) controls:
 // [[nodiscard]] attributes on STL functions
 
+// Determine if we should use [[msvc::known_semantics]] to communicate to the compiler
+// that certain type trait specializations have the standard-mandated semantics
+#ifndef __has_cpp_attribute
+#define _MSVC_KNOWN_SEMANTICS
+#elif __has_cpp_attribute(msvc::known_semantics)
+#define _MSVC_KNOWN_SEMANTICS [[msvc::known_semantics]]
+#else
+#define _MSVC_KNOWN_SEMANTICS
+#endif

```

Comment:
Wouldn't this be simpler?
```suggestion
#if defined(__has_cpp_attribute) && __has_cpp_attribute(msvc::known_semantics)
#define _MSVC_KNOWN_SEMANTICS [[msvc::known_semantics]]
#else
#define _MSVC_KNOWN_SEMANTICS
#endif
```

Reply:
Unfortunately, that doesn't work:

```
C:\Temp>type meow.cpp
#ifndef UNKNOWN_MACRO_1
#define _MSVC_KNOWN_SEMANTICS
#elif UNKNOWN_MACRO_1(msvc::known_semantics)
#define _MSVC_KNOWN_SEMANTICS [[msvc::known_semantics]]
#else
#define _MSVC_KNOWN_SEMANTICS
#endif

#if defined(UNKNOWN_MACRO_2) && UNKNOWN_MACRO_2(msvc::known_semantics)
#define _MSVC_KNOWN_SEMANTICS [[msvc::known_semantics]]
#else
#define _MSVC_KNOWN_SEMANTICS
#endif

C:\Temp>cl /EHsc /nologo /W4 /c meow.cpp
meow.cpp
meow.cpp(9): warning C4067: unexpected tokens following preprocessor directive - expected a newline

C:\Temp>clang-cl -m32 /EHsc /nologo /W4 /c meow.cpp
meow.cpp(9,33): error: function-like macro 'UNKNOWN_MACRO_2' is not defined
#if defined(UNKNOWN_MACRO_2) && UNKNOWN_MACRO_2(msvc::known_semantics)
                                ^
1 error generated.
```

Reply:
Oh, I thought it would work.

Reply:
This is the traditional (clang had `__has_cpp_attribute` as an extension for years before C++17 standardized it) form for `__has_cpp_attribute` for code which must be portable to compilers that don't have `__has_cpp_attribute`. (Since `__has_cpp_attribute` is a reserved name, we don't know that `__has_cpp_attribute(meow)` in `defined(__has_cpp_attribute) && __has_cpp_attribute(msvc::known_semantics)` won't blow up the compile.)

We currently support at least one front-end (NVCC) which may not support `__has_cpp_attribute` (I don't recall). While I doubt that it explodes when it sees `__has_cpp_attribute(meow)`, it's simple enough to just use the idiomatic form and not worry about it.

Reply:
C++ is fun!
* `&&` short-circuits!
* ... except when it has been overloaded
* ... and in the preprocessor
* ... and in metaprogramming (which is why we use `conjunction`)
* ... except that it does short-circuit in concepts! (I think).

Thanks for the suggestion - sometimes things do work and nobody has thought of them yet, so better to suggest than remain quiet. 😸

Reply:
Thanks for explanation

Reply:
> C++ is fun!
> 
> * `&&` short-circuits!
> * ... except when it has been overloaded
> * ... and in the preprocessor

The thing that `&&` short-circuits is evaluation; not macro-expansion, 

> * ... and in metaprogramming (which is why we use `conjunction`)

nor instantiation -

> * ... except that it does short-circuit in concepts! (I think).

but it _does_ short-circuit substitution of template parameters into atomic constraints when evaluating the associated constraints of a declaration. Which means that hoops we must jump through to avoid instantiation after failure in C++17:
```c++
// Don't instantiate other_trait<U> if some_trait<T> is false or invalid
template <class T, class U, class = enable_if_t<some_trait<T>>, class = enable_if_t<other_trait<U>>>
void meow() {}
```
is _much_ simpler in C++20:
```c++
// Don't instantiate other_concept<U> if some_concept<T> is false
template <some_concept T, other_concept U>
void meow() {}
```